semgrep-core/src/matching/SubAST_generic.ml

```diff
@@ -93,8 +93,11 @@ let subexprs_of_stmt_kind = function
 let subexprs_of_stmt st = subexprs_of_stmt_kind st.s
 
 (* used for deep expression matching *)
-let subexprs_of_expr e =
+let subexprs_of_expr with_symbolic_propagation e =
   match e.e with
+  | N (Id (_, { id_svalue = { contents = Some (Sym e1) }; _ }))
```

Comment:
Shouldn't we allow also when id_svalue is a Lit?

Reply:
Actually do we still need Lit now that it can be represented by Sym (Lit ...)?
Or it's still nice to have this special case encoded specially (but then we need to add the invariant somewhere
that we never can have Sym (L ...)

Reply:
Good point, that invariant is supposed to hold (you never have `Sym (Lit ...)` so it should be documented and maybe even checked at some places.

We could add `Lit` here too but I believe it is not needed. This is because if `x` equals to `1`,  then as long as `x` is enumerated by `subexprs_of_expr` we are fine.  You can always match `1` against `x` (if `x` is known to be equals to `1`). The problem is when `x` is equals to `!y`, and the pattern is `<... y ...>`. In that case, if only `x` is enumerated,  then `y` cannot be matched directly against `!y`. We need to say that `!y` is a subexpression of `x` so that a recursive call on `!y` will finally yield `y`.